src/drivers/fluid_midikit2.cpp

```diff
@@ -0,0 +1,161 @@
+            }
+            if (event)
+            {
+                callback->driver.handler(callback->driver.data, event);
+            }
+        }
+    }
+public:
+    fluid_midikit2_driver_t* callback;
+};
+
+void fluid_midikit2_driver_settings(fluid_settings_t *settings)
+{
+    fluid_settings_register_str(settings, "midi.midikit2.id", "pid", 0);
+}
+
+static void fluid_coremidi_autoconnect(fluid_midikit2_driver_t *dev)
+{
+    int32 id = 0;
+    BMidiProducer* producer = NULL;
+    while ((producer = BMidiRoster::NextProducer(&id)) != NULL)
+    {
+        producer->Connect(dev->m_consumer);
+        producer->Release();
+    }
+}
+
+fluid_midi_driver_t*
+new_fluid_midikit2_driver(fluid_settings_t *settings, handle_midi_event_func_t handler, void *data)
+{
+    fluid_midikit2_driver_t* driver = NULL;
+    char* id = NULL;
+    char clientid[128];
+
+    memset(clientid, 0, sizeof(clientid));
+    driver = FLUID_MALLOC(sizeof(fluid_midikit2_driver_t));
+
+    if(driver == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "Out of memory");
+        return NULL;
+    }
+
+    driver->driver.handler = handler;
+    driver->driver.data = data;
+
+    driver->parser = new_fluid_midi_parser();
+
+    if(driver->parser == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "Out of memory");
+        FLUID_FREE(driver);
+        return NULL;
+    }
+
+    fluid_settings_dupstr(settings, "midi.midikit2.id", &id);
+    if(id != NULL)
+    {
+        if(FLUID_STRCMP(id, "pid") == 0)
+        {
+            FLUID_SNPRINTF(clientid, sizeof(clientid), "FluidSynth virtual port (%d)", getpid());
+        }
+        else
+        {
+            FLUID_SNPRINTF(clientid, sizeof(clientid), "FluidSynth virtual port (%s)", id);
+        }
+
+        FLUID_FREE(id);   /* -- free id string */
+    }
+    else strncpy(clientid, "FluidSynth virtual port", sizeof(clientid));
+
+    driver->m_consumer = new fluid_midikit2_input;
+    if(!driver->m_consumer->IsValid())
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to create the MIDI consumer");
+        driver->m_consumer->Release();
+        delete_fluid_midi_parser(driver->parser);
+        FLUID_FREE(driver);
+        return NULL;
+    }
+
+    ((fluid_midikit2_input*)driver->m_consumer)->callback = driver;
+    driver->m_consumer->SetName(clientid);
+    driver->m_consumer->Register();
+    fluid_settings_getint(settings, "midi.autoconnect", &driver->autoconn_inputs);
+
+    if (driver->autoconn_inputs)
+    {
+        fluid_coremidi_autoconnect(driver);
+    }
+
+    return (fluid_midi_driver_t *)driver;
+}
+
+void
+delete_fluid_midikit2_driver(fluid_midi_driver_t *p)
+{
+    fluid_return_if_fail(p != NULL);
+    if (p)
+    {
+        fluid_midikit2_driver_t* driver = (fluid_midikit2_driver_t*)p;
+        driver->m_consumer->Unregister();
+        driver->m_consumer->Release();
+        delete_fluid_midi_parser(driver->parser);
+        FLUID_FREE(driver);
```

Comment:
You're leaking `driver->m_consumer`.

Reply:
Release() will delete it by itself as [it's ref-counted](https://www.haiku-os.org/docs/api/group__midi2.html#book_midi2refcount).

Reply:
Meanwhile I discovered the implementation of BMidiEndpoint: https://github.com/haiku/haiku/blob/c8252422b8b3a2a293fb3be039eec6f87b30b467/src/kits/midi2/MidiEndpoint.cpp#L118-L135

So it turns out that they do make a hard call to `delete this`. If you used `new[]` to allocate the endpoint this will cause undefined behavior. And if you stack allocated this you might get a log message: https://github.com/haiku/haiku/blob/c8252422b8b3a2a293fb3be039eec6f87b30b467/src/kits/midi2/MidiEndpoint.cpp#L258-L260

Poor implementation, IMO. But still: You must declare an empty destructor for `fluid_midikit2_input` and mark it `virtual` to make this cleanup work as intended.